rotkehlchen/chain/solana/utils.py

```diff
@@ -142,6 +165,63 @@ def get_metadata_account(token_address: SolanaAddress, metadata_program: Pubkey)
     )[0]  # return only address from tuple[address, nonce]
 
 
+def is_token_nft(
+        token_address: SolanaAddress,
+        mint_info: MintInfo,
+        metadata: MetadataInfo,
+) -> bool:
+    """Determine if a solana token is an NFT using the provided mint info and metadata.
+
+    Uses the following heuristics to determine if the token is an NFT:
+    - the is_nft field in the metadata if present (from the metaplex token standard)
+    - if the decimals are not zero then it is not an NFT
+    - if the decimals are zero and supply is one then it is an NFT
+    - if decimals are zero but supply is greater than one, check if the offchain metadata
+      looks like a semi-fungible NFT.
+
+    Returns a boolean indicating if the token is an NFT.
+    """
+    if metadata.is_nft is not None:
+        return metadata.is_nft
+    if mint_info.decimals != 0:  # decimals > 0 is always a normal fungible token
+        return False
+    if mint_info.supply == 1:  # decimals = 0 and supply = 1 is always an NFT
+        return True
+
+    # decimals == 0 and supply > 1 - could be either a semi-fungible (classed as NFT) or
+    # simply a normal fungible token with zero decimals (rare but possible)
+    # Check if the offchain metadata contains NFT specific data.
+    if (offchain_metadata := query_offchain_metadata(metadata.uri)) is None:
+        # Since normal fungible tokens with zero decimals are quite rare, assume it's an NFT
+        # if the offchain metadata is unavailable.
+        log.warning(
+            f'Offchain metadata is unavailable for solana token {token_address}. '
+            f'Assuming it is a semi-fungible NFT since it has 0 decimals and supply > 1.',
+        )
+        return True
+
+    if any(k in offchain_metadata for k in ('attributes', 'collection', 'properties')):
+        return True
+
+    log.debug(
+        f'Found solana token {token_address} with 0 decimals and supply > 1 but its '
+        f'offchain metadata does not look like a semi-fungible NFT. '
+        f'Classing it as a normal fungible token.',
+    )
+    return False
+
+
+def query_offchain_metadata(uri: str) -> dict[str, str] | None:
+    """Query the offchain metadata for the given token uri.
+    Returns the offchain metadata as a dict or None if the query fails.
+    """
+    try:
+        return request_get_dict(uri)
+    except (RemoteError, UnableToDecryptRemoteData) as e:
+        log.error(f'Failed to query offchain metadata uri {uri} due to {e}')
+        return None
```

Comment:
will this be used elsewhere other than in tests? it doesn't look like it should be a standalone function.